helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java

```diff
@@ -177,15 +176,26 @@ private void assignOrphans() {
     while (it.hasNext()) {
       Replica replica = it.next();
       boolean added = false;
-      int startIndex = computeRandomStartIndex(replica);
-      for (int index = startIndex; index < startIndex + _liveNodesList.size(); index++) {
-        Node receiver = _liveNodesList.get(index % _liveNodesList.size());
-        if (receiver.capacity > receiver.currentlyAssigned && receiver.canAdd(replica)) {
-          receiver.currentlyAssigned = receiver.currentlyAssigned + 1;
-          receiver.nonPreferred.add(replica);
-          receiver.newReplicas.add(replica);
-          added = true;
-          break;
+
+      // first find if it preferred node still has capacity
+      Node preferred = _preferredAssignment.get(replica);
+      if (preferred.capacity > preferred.currentlyAssigned && preferred.canAdd(replica)) {
```

Comment:
Can you refactor this into a method, since this is basically the same code as below?